client/src/Thumbnail.jsx

```diff
@@ -1,22 +1,18 @@
 import React, { useEffect, forwardRef } from "react";
 import PropTypes from "prop-types";
 
-const Thumbnail = forwardRef(({ scrollTo, paginator, selected, number }, ref) => {
-    const label = paginator.getLabel(number);
-    const selectPage = () => {
-        paginator.setPage(number);
-    };
+const Thumbnail = forwardRef(({ scrollTo, getMagicLabel, setPage, getImageUrl, selected, number }, ref) => {
     useEffect(() => {
         if (selected) {
             scrollTo(number);
         }
     }, [selected]);
 
     return (
-        <div onClick={selectPage} className={"thumbnail" + (selected ? " selected" : "")} ref={ref}>
-            <img alt="" src={paginator.getImageUrl(number, "thumb")} />
+        <div onClick={() => setPage(number)} className={"thumbnail" + (selected ? " selected" : "")} ref={ref}>
+            <img alt="" src={getImageUrl(number, "thumb")} />
             <div className="number">{number + 1}</div>
-            <div className={"label" + (null === paginator.getLabel(number, false) ? " magic" : "")}>{label}</div>
+            <div className={"label" + (!getMagicLabel(number) ? " magic" : "")}>{getMagicLabel(number)}</div>
```

Comment:
Here's an example of a possible problem with label use. I think what we might really want here is something like:

<pre>
            <div className={"label" + (null === getUserLabel(number) ? " magic" : "")}>{getBestLabel(number)}</div>
</pre>

In other words, display the most appropriate label for the thumbnail, but if it's an auto-generated magic label, apply an extra class so we can style it differently.